src/cluster.h

```diff
@@ -173,257 +51,102 @@ typedef struct {
-static_assert(offsetof(clusterMsg, mflags) == 2253, "unexpected field offset");
-static_assert(offsetof(clusterMsg, data) == 2256, "unexpected field offset");
-
-#define CLUSTERMSG_MIN_LEN (sizeof(clusterMsg)-sizeof(union clusterMsgData))
-
-/* Message flags better specify the packet content or are used to
- * provide some information about the node state. */
-#define CLUSTERMSG_FLAG0_PAUSED (1<<0) /* Master paused for manual failover. */
-#define CLUSTERMSG_FLAG0_FORCEACK (1<<1) /* Give ACK to AUTH_REQUEST even if
-                                            master is up. */
-#define CLUSTERMSG_FLAG0_EXT_DATA (1<<2) /* Message contains extension data */
-
 /* ---------------------- API exported outside cluster.c -------------------- */
+/*
+ * Functions requiring per-clustering mechanism implementation: Lifecycle events
+ */
 void clusterInit(void);
 void clusterInitListeners(void);
 void clusterCron(void);
 void clusterBeforeSleep(void);
-clusterNode *getNodeByQuery(client *c, struct redisCommand *cmd, robj **argv, int argc, int *hashslot, int *ask);
-int verifyClusterNodeId(const char *name, int length);
-clusterNode *clusterLookupNode(const char *name, int length);
-int clusterRedirectBlockedClientIfNeeded(client *c);
-void clusterRedirectClient(client *c, clusterNode *n, int hashslot, int error_code);
-void migrateCloseTimedoutSockets(void);
 int verifyClusterConfigWithData(void);
+void clusterUpdateMyselfFlags(void);
 unsigned long getClusterConnectionsCount(void);
+void clusterUpdateMyselfHostname(void);
+void clusterUpdateMyselfAnnouncedPorts(void);
+void clusterUpdateMyselfIp(void);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: cluster commands
+ */
+void clusterCommand(client *c);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: Pub/Sub and module messages
+ */
 int clusterSendModuleMessageToTarget(const char *target, uint64_t module_id, uint8_t type, const char *payload, uint32_t len);
 void clusterPropagatePublish(robj *channel, robj *message, int sharded);
+void slotToChannelAdd(sds channel);
+void slotToChannelDel(sds channel);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: Debug and info actions
+ */
+void freeThisNodesLink(clusterNode *node);
+void freeNodeInboundLink(clusterNode *node);
+sds clusterGenNodesDescription(client *c, int filter, int use_pport);
+sds genClusterInfoString(void);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: "Accessor" functions
+ */
+char* clusterNodeLastKnownIp(clusterNode *node);
+int clusterNodePort(clusterNode *node);
+clusterNode* clusterNodeGetSlaveof(clusterNode *node);
+int clusterNodeConfirmedReachable(clusterNode* node);
+int clusterNodeIsMaster(clusterNode* node);
+int clusterNodeIsSlave(clusterNode* node);
+int clusterNodeIsFailing(clusterNode* node);
+int clusterNodeTimedOut(clusterNode* node);
+int clusterNodeIsMyself(clusterNode* node);
```

Comment:
If this is a handle, this would make more sense to not be per-clustering, since you should be able to still do myself == node. It would be odd to have the node "change" silently underneath you. Ditto about getMyClusterNode.

Reply:
Actually, this function is used both in the "shared" cluster code (cluster.c) and in module.c (grrrr). Saying "myself == node" requires holding a static myself. Since IsMyself needs to be determined both within the clustering code and within redis (module.c), it makes more sense to me that the clustering hold the static myself and then answer questions about it.